File: challenge-8/dec_to_bin.c
/* C Program to convert input decimal number
* to binary and print it.
* Input Constraint: Decimal number > 7
*/

#include <stdio.h>
#include <math.h>

// Converts binary to dec
void dec_to_bin(int decimal_num);

// Driver code
int main()
{
	int decimal_num;

	// Prompt input until constraint satisfied
	do
	{
		printf("Input decimal num: ");
		scanf("%d", &decimal_num);
	} while (decimal_num < 8);

	dec_to_bin(decimal_num);

	return 0;
}

// Converts binary to dec
void dec_to_bin(int decimal_num)
{	
	// n bits are required to represent 2^n - 1 in decimal.
	// hence log base 2 gives the num of binary bits required
	// to represent decimal_num. Added 1 to complete range
	// For example, decimal_num = 64, log base(2) = 6 + 1 = 7
	// 7 bits required for writing 64 in binary
	int size = log2(decimal_num) + 1;
	int binary_num[size];

	int count = 0;

	// Iteratively Store each remainder when dividing by 2
	// as binary bits until decimal_num is 0
	while (decimal_num != 0)
	{
		binary_num[count] = decimal_num % 2;
		decimal_num /= 2;
		count++;
	}

	printf("Binary: ");

	// Print binary_num in reverse
	for (int i = size - 1; i >= 0; i--)
	{
		printf("%d", binary_num[i]);
	}

	printf("\n");
}

File: challenge-8/DecToBin.java
/* 
* Java program to convert a decimal number into a
* binary number and print it
* Input Constraint: decimal number > 7
*/

import java.util.Scanner;

public class DecToBin {

	// Convery decimal input to binary
	private static void decimalToBinary(int decimal_num) {

		// log2 N = log10 N / log10 2
		// Find the number of bits required to represent
		// var decimal_num in binary
		int size = 1 + (int) (Math.log(decimal_num) / Math.log(2));

		int[] binaryNum = new int[size];
		int count = 0;

		// Generate and store binary pattern by taking
		// remainder and divising by 2 each time
		while (decimal_num != 0) {
			binaryNum[count] = decimal_num % 2;
			decimal_num /= 2;
			count++;
		}

		System.out.print("Binary: ");

		// Print binaryNum in reverse
		for (int i = size - 1; i >= 0; i--) {
			System.out.print(binaryNum[i]);
		}

		// newline
		System.out.println();
	}

	// Driver code
	public static void main(String[] args) {

		// Scanner object for input
		Scanner scanner = new Scanner(System.in);

		int decimal_num;
		System.out.print("Input decimal num: ");

		// Prompt for input until constraint satisfied
		do {
			decimal_num = scanner.nextInt();
		} while (decimal_num < 8);

		decimalToBinary(decimal_num);
	}
}